src/mt.rs

```diff
@@ -0,0 +1,357 @@
+    }
+
+    /// Change this meta-tracer builder's `hot_threshold` value.
+    pub fn hot_threshold(mut self, hot_threshold: HotThreshold) -> Self {
+        self.hot_threshold = hot_threshold;
+        self
+    }
+}
+
+#[derive(Clone)]
+/// A meta-tracer. Note that this is conceptually a "front-end" to the actual meta-tracer akin to
+/// an `Rc`: this struct can be freely `clone()`d without duplicating the underlying meta-tracer.
+pub struct MT {
+    inner: Arc<MTInner>,
+}
+
+impl MT {
+    /// Return this meta-tracer's hot threshold.
+    pub fn hot_threshold(&self) -> HotThreshold {
+        self.inner.hot_threshold.load(Ordering::Relaxed)
+    }
+
+    /// Create a new thread that can be used in the meta-tracer: the new thread that is created is
+    /// handed a [`MTThread`](struct.MTThread.html) from which the `MT` itself can be accessed.
+    pub fn spawn<F, T>(&self, f: F) -> io::Result<JoinHandle<T>>
+    where
+        F: FnOnce(MTThread) -> T,
+        F: Send + UnwindSafe + 'static,
+        T: Send + 'static,
+    {
+        let mt_cl = self.clone();
+        thread::Builder::new().spawn(move || {
+            mt_cl.inner.active_threads.fetch_add(1, Ordering::Relaxed);
+            let r = catch_unwind(|| f(MTThreadInner::init(mt_cl.clone())));
+            mt_cl.inner.active_threads.fetch_sub(1, Ordering::Relaxed);
+            match r {
+                Ok(r) => r,
+                Err(e) => resume_unwind(e),
+            }
+        })
+    }
+}
+
+/// The innards of a meta-tracer.
+struct MTInner {
+    hot_threshold: AtomicU32,
+    active_threads: AtomicUsize,
+}
+
+/// It's only safe to have one `MT` instance active at a time.
+static MT_ACTIVE: AtomicBool = AtomicBool::new(false);
+
+impl MTInner {
+    /// Create a new `MT`, wrapped immediately in an [`MTThread`](struct.MTThread.html).
+    fn init(hot_threshold: HotThreshold) -> MTThread {
+        // A process can only have a single MT instance.
+
+        // In non-testing, we panic if the user calls this method while an MT instance is active.
+        #[cfg(not(test))]
+        {
+            if MT_ACTIVE.swap(true, Ordering::Relaxed) {
+                panic!("Only one MT can be active at once.");
+            }
+        }
+
+        // In testing, we simply sleep until the other MT instance has gone away: this has the
+        // effect of serialising tests which use MT (but allowing other tests to run in parallel).
+        #[cfg(test)]
+        {
+            loop {
+                if !MT_ACTIVE.swap(true, Ordering::Relaxed) {
+                    break;
+                }
+                thread::sleep(Duration::from_millis(10));
+            }
+        }
+
+        let mtc = Self {
+            hot_threshold: AtomicU32::new(hot_threshold),
+            active_threads: AtomicUsize::new(1),
+        };
+        let mt = MT {
+            inner: Arc::new(mtc),
+        };
+        MTThreadInner::init(mt)
+    }
+}
+
+impl Drop for MT {
```

Comment:
Should we move this next to the def of `MT`?

Reply:
Oops, yes!